smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -55,17 +60,520 @@
+        """Check if a database is reachable.
+
+        :returns: True if connection succeeds, False otherwise."""
+        try:
+            if self._client:
+                return await self._client.ping()
+        except redisex.ConnectionError:
+            logger.info(f"Cannot ping db {self._address}")
+
+        return False
+
+
+class DBMemoryCollector(DBCollector):
+    """A collector that collects memory usage information from
+    an orchestrator instance"""
+
+    def __init__(self, entity: JobEntity, sink: Sink) -> None:
+        super().__init__(entity, sink)
+        self._columns = ["used_memory", "used_memory_peak", "total_system_memory"]
+
+    async def post_prepare(self) -> None:
+        """Hook called after the db connection is established"""
+        await self._sink.save(
+            "timestamp",
+            "used_memory",
+            "used_memory_peak",
+            "total_system_memory",
+        )
+
+    async def collect(self) -> None:
+        if not self.enabled:
+            logger.debug(f"{type(self).__name__} is not enabled")
+            return
+
+        await self.prepare()
+        if not self._client:
+            logger.warning(f"{type(self).__name__} cannot collect")
+            return
+
+        try:
+            if not await self._check_db():
+                return
+
+            db_info = await self._client.info("memory")
+            value = (
+                self.timestamp(),
+                float(db_info["used_memory"]),
+                float(db_info["used_memory_peak"]),
+                float(db_info["total_system_memory"]),
+            )
+
+            await self._sink.save(*value)
+        except Exception as ex:
+            logger.warning(f"Collect failed for {type(self).__name__}", exc_info=ex)
+
+
+class DBConnectionCollector(DBCollector):
+    """A collector that collects client connection information from
+    an orchestrator instance"""
+
+    async def post_prepare(self) -> None:
+        """Hook called after the db connection is established"""
+        await self._sink.save("timestamp", "client_id", "address")
+
+    async def collect(self) -> None:
+        if not self.enabled:
+            logger.debug(f"{type(self).__name__} is not enabled")
+            return
+
+        await self.prepare()
+        if not self._client:
+            logger.warning(f"{type(self).__name__} is not connected and cannot collect")
+            return
+
+        now_ts = self.timestamp()  # ensure all results have the same timestamp
+
+        try:
+            if not await self._check_db():
+                return
+
+            clients = await self._client.client_list()
+
+            value = [(now_ts, item["id"], item["addr"]) for item in clients]
+
+            for client_info in value:
+                await self._sink.save(*client_info)
+        except Exception as ex:
+            logger.warning(f"Collect failed for {type(self).__name__}", exc_info=ex)
```

Comment:
Things are going to get a little weird when reporting the connection collectors. If I have an application that connects 128 clients to a single shard DB, what actually is going to reported is 128+[number of telemetry clients]. A user who doesn't know how the telemetry monitor works might go down a rabbit hole to figure out why they have more clients than expected. I'm beginning to wonder if we should share a Redis client (or maybe a pool of Redis clients) for the collectors and be able to report non-telemetry clients